File: app/src/main/java/wangdaye/com/geometricweather/db/controllers/HistoryEntityController.java
package wangdaye.com.geometricweather.db.controllers;

import android.annotation.SuppressLint;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

import wangdaye.com.geometricweather.common.basic.models.options.provider.WeatherSource;
import wangdaye.com.geometricweather.db.GeometricWeatherDatabase;
import wangdaye.com.geometricweather.db.converters.WeatherSourceConverter;
import wangdaye.com.geometricweather.db.entities.HistoryEntity;

public class HistoryEntityController extends AbsEntityController {

    public static void insertHistoryEntity(@NonNull GeometricWeatherDatabase database,
                                           @NonNull HistoryEntity entity) {
        database.historyDao().insertHistoryEntity(entity);
    }

    public static void deleteLocationHistoryEntity(@NonNull GeometricWeatherDatabase database,
                                                   @NonNull String cityId,
                                                   @NonNull WeatherSource source) {
        database.historyDao().deleteLocationHistoryEntity(
                cityId,
                new WeatherSourceConverter().convertToDatabaseValue(source)
        );
    }

    @SuppressLint("SimpleDateFormat")
    @Nullable
    public static HistoryEntity selectYesterdayHistoryEntity(
            @NonNull GeometricWeatherDatabase database,
            @NonNull String cityId,
            @NonNull WeatherSource source,
            @NonNull Date currentDate) {
        try {
            SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd");
            Date today = format.parse(format.format(currentDate));
            if (today == null) {
                throw new NullPointerException("Get null Date object.");
            }

            Calendar calendar = Calendar.getInstance();
            calendar.setTime(today);
            calendar.add(Calendar.DATE, -1);
            Date yesterday = calendar.getTime();

            return database.historyDao().selectHistoryEntity(
                    cityId,
                    new WeatherSourceConverter().convertToDatabaseValue(source),
                    yesterday,
                    today
            );
        } catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    @NonNull
    public static List<HistoryEntity> selectHistoryEntityList(
            @NonNull GeometricWeatherDatabase database,
            @NonNull String cityId,
            @NonNull WeatherSource source) {
        return getNonNullList(
                database.historyDao().selectHistoryEntityList(
                        cityId,
                        new WeatherSourceConverter().convertToDatabaseValue(source)
                )
        );
    }
}

File: data/src/main/java/wangdaye/com/geometricweather/db/controllers/HistoryEntityController.kt
package wangdaye.com.geometricweather.db.controllers

import android.annotation.SuppressLint
import wangdaye.com.geometricweather.common.basic.models.options.provider.WeatherSource
import wangdaye.com.geometricweather.db.GeometricWeatherDatabase
import wangdaye.com.geometricweather.db.converters.WeatherSourceConverter
import wangdaye.com.geometricweather.db.entities.HistoryEntity
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date

object HistoryEntityController {

    @JvmStatic
    fun insertHistoryEntity(database: GeometricWeatherDatabase, entity: HistoryEntity) {
        database.historyDao().insertHistoryEntity(entity)
    }

    @JvmStatic
    fun deleteLocationHistoryEntity(
        database: GeometricWeatherDatabase,
        cityId: String,
        source: WeatherSource
    ) {
        database.historyDao().deleteLocationHistoryEntity(
            cityId,
            WeatherSourceConverter().convertToDatabaseValue(source)
        )
    }

    @SuppressLint("SimpleDateFormat")
    @JvmStatic
    fun selectYesterdayHistoryEntity(
        database: GeometricWeatherDatabase,
        cityId: String,
        source: WeatherSource,
        currentDate: Date
    ): HistoryEntity? {
        return try {
            val format = SimpleDateFormat("yyyy-MM-dd")
            val today = format.parse(format.format(currentDate))
                ?: throw NullPointerException("Get null Date object.")

            val calendar = Calendar.getInstance()
            calendar.time = today
            calendar.add(Calendar.DATE, -1)
            val yesterday = calendar.time

            database.historyDao().selectHistoryEntity(
                cityId,
                WeatherSourceConverter().convertToDatabaseValue(source),
                yesterday,
                today
            )
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    @JvmStatic
    fun selectHistoryEntityList(
        database: GeometricWeatherDatabase,
        cityId: String,
        source: WeatherSource
    ): List<HistoryEntity> {
        return getNonNullList(
            database.historyDao().selectHistoryEntityList(
                cityId,
                WeatherSourceConverter().convertToDatabaseValue(source)
            )
        )
    }
}
